File: laba3/task2/C++/queue.cpp
#include <iostream>
#include "structures.h"

using namespace std;

void Queue::remFrontQueue()
{
	const Node* temp = this->head;
	this->head = this->head -> next;
	delete temp;
	
	if (this->head == nullptr) this->tail = nullptr;
}

void Queue::addBackQueue(const string& data)
{
    Node* node = createNodeQueue(data);
    if (this->head == nullptr){
    	this->head = node;
    	this->tail = node;
    	return;
    }
    this->tail->next = node;
    this->tail = node;
}

void Queue::printQueue() const
{
    const Node* curr = this->head;
    while (curr != nullptr) {
        cout << curr->data << " ";
        curr = curr->next;
    }
    cout << endl;
}

File: laba3/task3/Golang/queue.go
package main
import "fmt"

type Queue struct {
	head *Node
	tail *Node
}

func (this *Queue) createNodeQueue(data string) *Node {
	return createNode(data)
}

func (this *Queue) remFrontQueue() {
	if this.head == nil {
		return
	}
	// temp := this.head
	this.head = this.head.next
	
	if this.head == nil {
		this.tail = nil
	}
}

func (this *Queue) addBackQueue(data string) {
	node := this.createNodeQueue(data)
	if this.head == nil {
		this.head = node
		this.tail = node
		return
	}
	this.tail.next = node
	this.tail = node
}

func (this *Queue) printQueue() {
	curr := this.head
	for curr != nil {
		fmt.Print(curr.data, " ")
		curr = curr.next
	}
	fmt.Println()
}

func (this *Queue) getHead() *Node { return this.head }
func (this *Queue) getTail() *Node { return this.tail }